### DISCUSSION TITLE
ch7/ch7-13

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#197] [General]
# ch7/ch7-13

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch7/ch7-13.html

**Comment by Zacharia2** [comment]
Go 语言里接口的两种玩法

**第一种方式：关注“行为”，不在乎“你是谁”**

这是最常见、最符合直觉的用法。

想象一下，你有一个 `USB` 接口。一个 U 盘、一个移动硬盘、一个键盘，它们都能插到 USB 口上。
你的电脑（代码）只关心一件事：这个设备能不能“读数据”（`Read`）和“写数据”（`Write`）。它不关心插进来的是金士顿的 U 盘还是西数的移动硬盘。`io.Reader` 和 `io.Writer` 就是这个道理，任何东西，只要有 `Read` 方法，我就能当“读取器”用；只要有 `Write` 方法，我就能当“写入器”用。

**核心思想**：接口定义了一套“行为规范”（方法），只要满足这套规范的任何类型，我都能用。我关心的是你能“做什么”，而不是你“是什么”。这实现了代码的解耦和通用性。

---

**第二种方式：关注“身份”，用接口当“收纳盒”**

这种用法比较特别，重点不在于接口有什么方法，而在于它能“装”各种不同的东西。
想象一下，你有一个 `any` 类型的变量（在 Go 1.18 之前就是 `interface{}`）。这个变量就像一个万能收纳盒，什么都能往里放：可以放数字 `123`，可以放字符串 `"hello"`，也可以放一个结构体。
这时候，接口本身的方法（通常它没有方法）不重要了，重要的是它“能装东西”这个特性。当你想用这个盒子里的东西时，你得先“打开盒子看看是什么”，这就是**类型断言**（`value, ok := i.(string)`）。
作者把这个比喻为“可辨识联合”，意思就是这个接口值代表了一个“类型的集合”，你需要用类型断言来“辨识”出它当前到底是哪个具体类型。

**核心思想**：接口被用作一个“容器”，用来统一处理多种不同类型的值。代码的逻辑分支是根据运行时“装”进去的具体类型来决定的。

---

**简单总结一下**

*   **第一种（行为多态）**：**“你只要有能耐，我不管你是谁。”**
    *   例子：`io.Writer` 接口，不管你是文件、网络连接还是内存缓冲区，只要有 `Write` 方法，我就能往里写。
    *   重点：**方法**。
*   **第二种（类型联合）**：**“来者不拒，都先进这个屋。待会儿我再挨个问你们是谁，然后区别对待。”**
    *   例子：函数接收 `any` 类型参数，内部用 `switch v := i.(type)` 来判断是 `int`、`string` 还是别的什么，然后执行不同逻辑。
    *   重点：**具体类型**。
作者提到，你如果懂 OOP，可以把第一种理解为“子类型多态”，第二种理解为“非参数多态”。但就像他说的，不用记这些名词，理解上面两种用法就行。
而且，他特别说明，接下来要讲的都是**第二种**用法的例子，也就是把接口当“收纳盒”用的情况。这样是不是清晰多了？

--------------------------------
以上内容由AI生成，仅供参考和借鉴

**Reply by ikun-1** [reply]
清晰
通透
